### PR TITLE
[`core`] Refactor of `gradient_checkpointing`

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Alternative to https://github.com/huggingface/transformers/pull/26917 

This way we make `set_gradient_checkpointing` more modulable, as requested by some users - e.g. https://github.com/huggingface/transformers/issues/21381#issuecomment-1757690624 

Fixes some issues with DDP such as: https://github.com/huggingface/trl/issues/835 

Also removed GC support from `TFSwin` as in theory `gradient_checkpointing` is used only for PT models.

Added also a CI tests for that

For users that want to use `gradient_checkpointing` with `use_reentrant=False`:

```python
...
model.enable_gradient_checkpointing(gradient_checkpointing_kwargs={"use_reentrant": False})
```

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -1845,7 +1858,7 @@ def gradient_checkpointing_disable(self):
         activations".
         """
         if self.supports_gradient_checkpointing:
-            self.apply(partial(self._set_gradient_checkpointing, value=False))
+            self.apply(partial(self._set_gradient_checkpointing, gradient_checkpointing_func=None))
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:1861]: WHen we disable gradient checkpointing, I think the `module.gradient_checkpointing` will still be `True`. 
Let's make `module.gradient_checkpointing` into a property to be sure we always check if the function is none or not WDYT?

**Review by ArthurZucker — COMMENTED** [review]
Very nice cleanup!

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/modeling_utils.py:1861]: Property could go at the ModelMixin level ?

**Comment by ArthurZucker** [comment]
Can you add a test to make sure setting and unsetting both work as expected (specifically for the fix we are implementing in TRL)

**Comment by LysandreJik** [comment]
+1

File: src/transformers/models/blip_2/modeling_blip_2.py [hunk 2]
```diff
@@ -297,9 +297,14 @@ def _init_weights(self, module):
...
-        if isinstance(module, Blip2Encoder):
-            module.gradient_checkpointing = value
+    def _set_gradient_checkpointing(self, module, gradient_checkpointing_func=None):
+        if isinstance(module, (Blip2Encoder, Blip2QFormerEncoder)):
+            module.gradient_checkpointing_func = gradient_checkpointing_func
+            module.gradient_checkpointing = gradient_checkpointing_func is not None
+
+        # Enable / disable GC for the language model as well
+        if hasattr(self, "language_model") and hasattr(self.language_model, "_set_gradient_checkpointing"):
+            self.language_model._set_gradient_checkpointing(module, gradient_checkpointing_func)
```
Comment (younesbelkada) [inline src/transformers/models/blip_2/modeling_blip_2.py:307]: BLIP2 never propagated `gradient_checkpointing` to its `language_model`

File: src/transformers/models/instructblip/modeling_instructblip.py [hunk 3]
```diff
@@ -304,9 +304,14 @@ def _init_weights(self, module):
...
-        if isinstance(module, InstructBlipEncoder):
-            module.gradient_checkpointing = value
+    def _set_gradient_checkpointing(self, module, gradient_checkpointing_func=None):
+        if isinstance(module, (InstructBlipEncoder, InstructBlipQFormerEncoder)):
+            module.gradient_checkpointing_func = gradient_checkpointing_func
+            module.gradient_checkpointing = gradient_checkpointing_func is not None
+
+        # Enable / disable GC for the language model as well
+        if hasattr(self, "language_model") and hasattr(self.language_model, "_set_gradient_checkpointing"):
+            self.language_model._set_gradient_checkpointing(module, gradient_checkpointing_func)
```
Comment (younesbelkada) [inline src/transformers/models/instructblip/modeling_instructblip.py:314]: Same here

File: src/transformers/models/vitmatte/modeling_vitmatte.py [hunk 4]
```diff
@@ -86,9 +86,15 @@ def _init_weights(self, module):
...
+    def _set_gradient_checkpointing(self, module, gradient_checkpointing_func=None):
         if isinstance(module, BackboneMixin):
-            module.gradient_checkpointing = value
+            module.gradient_checkpointing_func = gradient_checkpointing_func
+            module.gradient_checkpointing = gradient_checkpointing_func is not None
+
+            for backbone_module in module.modules():
+                if hasattr(backbone_module, "gradient_checkpointing"):
+                    backbone_module.gradient_checkpointing_func = gradient_checkpointing_func
+                    backbone_module.gradient_checkpointing = gradient_checkpointing_func is not None
```
Comment (younesbelkada) [inline src/transformers/models/vitmatte/modeling_vitmatte.py:97]: Another edge case here where the backbone has some modules that support GC but that attribute never being propagated

**Review by younesbelkada — COMMENTED** [review]
It turns out ~30 architectures were not properly using `gradient_checkpointing`, I left 3 comments to be aware of

**Review by ArthurZucker — COMMENTED** [review]
I think we should use the call rather than forward to have the hooks!

File: src/transformers/models/align/modeling_align.py [hunk 5]
```diff
@@ -1095,20 +1095,15 @@ def forward(
...
-                def create_custom_forward(module):
-                    def custom_forward(*inputs):
-                        return module(*inputs, past_key_value, output_attentions)
-
-                    return custom_forward
-
-                layer_outputs = torch.utils.checkpoint.checkpoint(
-                    create_custom_forward(layer_module),
+                layer_outputs = self.gradient_checkpointing_func(
+                    layer_module.__call__,
```
Comment (ArthurZucker) [inline src/transformers/models/align/modeling_align.py:1099]: let's document this in the gradient checkpointing doc (IMO important to know! why forward and call are different)

**Review by ArthurZucker — APPROVED** [review]
Thanks a lot, very nice cleanup! 🔥

**Comment by younesbelkada** [comment]
Ran some training tests with PEFT + GC using this branch and everything seem to pass! Merging once the CI is green

**Comment by lucasjinreal** [comment]
Whatis the difference of enable_gradient_checkpointing and gradient_checkpointing_enable??

**Comment by younesbelkada** [comment]
@lucasjinreal I can only see `gradient_checkpointing_enable` here: https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_utils.py

**Comment by younesbelkada** [comment]
🤯 
I think this was a typo, it is weird that you don't get any error, you should use `gradient_checkpointing_enable` https://github.com/huggingface/transformers/blob/af4c02622bfb4521367c459c6743014ef9be788d/src/transformers/modeling_utils.py#L2195
